### PR TITLE
FAD-5307 Role-based access control for routes and nav

### PR DESCRIPTION
**Pull request by jasonrhodes**

### CODE REVIEW

File: src/components/ProtectedRoute.js [hunk 1]
```diff
@@ -1,23 +0,0 @@
-import React, { Component } from 'react';
```
Comment (jasonrhodes) [inline src/components/ProtectedRoute.js]: This just moved to src/components/auth/ ... not sure why git didn't pick that up

File: src/components/auth/ProtectedRoute.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+export class ProtectedRoute extends Component {
+
+  renderComponent(propsFromRoute) {
+    const { component: Component, grants = [], condition } = this.props;
+
+    if (grants.length === 0) {
+      // TODO: make this render the loading page somehow??
+      return null;
+    }
```
Comment (jasonrhodes) [inline src/components/auth/ProtectedRoute.js]: if we don't do this check on `grants.length`, the component first mounts before we have the grants which means they won't have access to any page that requires any grants check, and then the redirect logic kicks in and sends them to `/dashboard` before we can actually figure out if they're allowed to see this route.

i'm all ears if there's a better way to do this! it was flashing an all white screen for a second or so before so i thought we could load that generic loading screen in that case, but now that i've moved the grants into the cookie it's pretty fast.

**Comment by jasonrhodes** [comment]
So to help with this review:

* The initial discussion is #98 
* I included docs for how most of this works: https://github.com/SparkPost/2web2ui/blob/hotfix/FAD-5307/docs/access-control.md
* Here are some notes about how the grants stuff is working:

We get grants from `/authenticate/grants` on every login, but the `<ProtectedRoute>` component has to wait until we have those grants before it can decide how to render/redirect. That was causing a 1 second flash of white screen, so I opted to store the grants in the auth cookie like we do in the current app. Getting them out of the cookie is very fast so things feel pretty snappy.

One problem I ran into with putting the grants in the cookie: I am storing the grants as an array of objects in the Redux store, each object has a key, a label, and a description. Trying to store that whole thing in the cookie resulted in a silent failure to write that cookie, I'm assuming because it was too large. So for now I'm storing the grants as a list of keys only, but instead of re-writing everything else to assume we only ever have the keys, if I find the grants in the cookie I store them mapped as objects with just the `key` property. 

tl;dr there's lots of map and unmap happening that we could probably clean up one way or another.

File: src/components/auth/ProtectedRoute.js (hunk 2 above)
Comment (jonambas) [reply src/components/auth/ProtectedRoute.js]: `return <Layout.Form loading={true} />`

or `<Layout.App>` to show the nav - they're logged and inside the app when this hits right?

File: src/components/navigation/Navigation.js [hunk 3]
```diff
@@ -14,8 +15,13 @@ class Navigation extends Component {
     open: false
   }
 
+  constructor(props) {
+    super(props);
+    this.state.navItems = prepareNavItems(this.props.store);
```
Comment (jonambas) [inline src/components/navigation/Navigation.js]: So `Layout` (and `Navigation`)  is re-mounted on every route, so this is going to run on every route change. Wonder if theres a way to do this once on log in.

Or do you think this isn't expensive?

File: src/config/default.js [hunk 4]
```diff
@@ -30,7 +30,10 @@ const config = {
     { time: 60 * 24 * 190, value: 'week', format: 'MMM Do' },
     { time: Infinity, value: 'month', format: 'MMM YY' }
   ],
+  showThingOnDash: true,
+  showSummaryReport: true,
```
Comment (jonambas) [inline src/config/default.js]: Are these just example flags?

File: src/App.js [hunk 5]
```diff
@@ -1,58 +1,31 @@
...
-    </div>
-  </Router>
-);
+export default function App() {
+  return (
+    <Router>
+      <div>
+        <AuthenticationGate />
+
+        {routes.map((route) => {
```
Comment (tough-griff) [inline src/App.js]: If we're not using nested routes in the `routes` config, we could wrap this collection with a [`<Switch>`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/Switch.md) to prevent a bunch of `<!-- react-emtpy: * -->`s.

File: src/App.js [hunk 6]
```diff
@@ -1,58 +1,31 @@
...
+          const MyRoute = route.public ? Route : ProtectedRoute;
+          route.exact = !(route.exact === false); // this makes exact default to true
+
+          if (route.redirect) {
+            return <Route key={route.path} {...route} render={() => <Redirect to={route.redirect} />} />;
```
Comment (tough-griff) [inline src/App.js]: `<Switch>` also allows us to use a nice shorthand here:
```jsx
<Redirect from={route.path} to={route.redirect} />
```

File: docs/access-control.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  }
+]
+```
+
+_Note: Routes default to exact: true because [it makes more sense](https://github.com/ReactTraining/react-router/issues/4958)._
+
+
+## Condition helpers
+
+Condition helpers are functions that return a condition function. Any time you need to use a condition, create a helper for that condition scenario. Whenever that returned function is called, it will be passed an object with keys `store` (the entire Redux store) and `config` (the current loaded tenant config), and it should return a boolean that reflects the access decision.
```
Comment (tough-griff) [inline docs/access-control.md]: `store` here might better be referred to as `state`, since we are getting the current state tree, not the Redux `store` object.

It's picky, but the `store` object itself contains more than just the state, and would require `getState()` calls.

File: src/components/auth/AccessControl.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    const show = condition({ store, config });
+
+    if (redirect && !show) {
+      return <Redirect to={redirect} />;
+    }
+    return show ? children : null;
+  }
+}
+
+export default connect((store) => ({ store }))(AccessControl);
```
Comment (tough-griff) [inline src/components/auth/AccessControl.js]: The component doesn't need to be aware of all state here:
```jsx
export const AccessControl = ({ children, redirect, show }) => {
  if (redirect && !show) return <Redirect to={redirect} />;
  return show ? children : null;
}

const mapStateToProps = (state, { condition = accept }) => ({
  show: condition({ state, config }),
});

export default connect(mapStateToProps)(AccessControl);
```

This could help perf by preventing rerenders on prop changes that the route doesn't care about.

File: src/components/auth/ProtectedRoute.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      )}/>
+    );
+  }
+}
+
+const mapStateToProps = ({ auth, currentUser = {}}) => ({
+  auth,
+  grants: currentUser.grants || []
+});
+export default withRouter(connect(mapStateToProps)(ProtectedRoute));
```
Comment (tough-griff) [inline src/components/auth/ProtectedRoute.js]: I think `withRouter` here is redundant, the `Route` component will pass the relevant props to it's children.

File: src/components/auth/ProtectedRoute.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        <Component {...propsFromRoute} />
+      </AccessControl>
+    );
+  }
+
+  render() {
+    const { component, auth, condition, ...rest } = this.props;
+    return (
+      <Route {...rest} render={(props) => (
+        auth.loggedIn ? this.renderComponent(props) : (
```
Comment (tough-griff) [inline src/components/auth/ProtectedRoute.js]: does `AuthenticationGate` take care of this logic for us? is it necessary to check logged in status in each route?

File: src/components/auth/ProtectedRoute.js (hunk 2 above)
Comment (jasonrhodes) [reply src/components/auth/ProtectedRoute.js]: Yes they're logged in, that sounds good to me (although will it be nav + blank page?)

File: src/components/navigation/Navigation.js (hunk 3 above)
Comment (jasonrhodes) [reply src/components/navigation/Navigation.js]: We could memoize some of it but that could get tricky… see my reply to Griffin's comment about moving Nav outside

File: src/config/default.js (hunk 4 above)
Comment (jasonrhodes) [reply src/config/default.js]: Yeah, I'll remove before merge.

**Comment by jasonrhodes** [comment]
@tough-griff since GitHub doesn't let me directly reply to general comments 😠 , this is re:
>I think wrapping the router in Layout, which could provide the navigation components outside of each individual route could help with loading/race condition issues, as well as prevent recomputing the nav component on each page. It might take some refactoring, but since that content is present on all pages it shouldn't need to be re-rendered all the time.

If we can figure out how to do this it seems fine to me. One thing I've been thinking about is how to handle dynamic conditions in all of this access control logic, so if a condition is something like:
```js
// helpers/conditions/accountStatus.js
export default function(status) {
  return ({ state }) => state.account.status === status
}

// config/routes.js
{
  path: /new/cool/page,
  condition: accountStatus('active')
}
```
What happens if an account is suspended while they're in the app? There are places where we refresh the account but if their account status updates, it'd be nice to re-render the routes and nav so that the conditions are re-checked. But having those run on every state change is, I agree, not a great solution to that.

It's not a huge deal that we solve this right now because we don't have conditions that need to be this dynamic, but I'd hate to build a system that can't handle dynamic conditions. If anyone has thoughts on how we could get this behavior, let me know!

tl;dr @jonambas if you think we can get the nav stuff to render outside of all the regular layouts, I think that'd probably be good for rendering/perf etc.

File: docs/access-control.md (hunk 7 above)
Comment (jasonrhodes) [reply docs/access-control.md]: This makes sense. I'm weirded out by calling it `state` bc then you technically have `this.props.state` which is super wacky to me ... could maybe use `globalState` but if `state` is a convention I can be convinced to get over it.

**Comment by jonambas** [comment]
re: @jasonrhodes 
Yeah, I think it'll take a bit of rework.

- We'd lose the `<Layout loading>` prop on all the pages, and would have to replace with the `<Loading>` component. 
- Would need to figure out how to load the right layout by route. Most protected routes would use `App` but there are a few that use `Form` - sign up flow, and I think a TOU page. Maybe we could pass it through routes.js:
```
{
  path: '/templates',
  component: templates.ListPage,
  layout: Layout.App
},
```

File: src/components/auth/ProtectedRoute.js (hunk 9 above)
Comment (jasonrhodes) [reply src/components/auth/ProtectedRoute.js]: Good call, I must've been doing some kind of `history.push` here at one point that's no longer around. Removing.

File: src/components/auth/ProtectedRoute.js (hunk 10 above)
Comment (jasonrhodes) [reply src/components/auth/ProtectedRoute.js]: Yeah not every route is protected, so we can't do that check in `<AuthenticationGate>` ... tbh I _think_ the only thing we even do in `<AuthenticationGate>` is initiate the check for the cookie and issue a login action if we find it, otherwise we could _probably_ handle the login/logout redirect stuff inside various chained actions, but I lost 4 or 5 days on that when I first set it up so I am a bit scared of messing with it at this point lol

File: src/components/auth/AccessControl.js (hunk 8 above)
Comment (jasonrhodes) [reply src/components/auth/AccessControl.js]: Interesting, I'm not sure if I follow how this will lead to less re-renders? It's arguably cleaner so I still kind of like it ... I guess `condition` will be run the same number of times but unless it produces a different bool value, the component won't be re-rendered?

Comment (tough-griff) [reply src/components/auth/AccessControl.js]: Yup! Passing the entire `state` object will trigger a lot of prop changes when the output would not otherwise change.

File: docs/access-control.md (hunk 7 above)
Comment (tough-griff) [reply docs/access-control.md]: As mentioned below, we can and probably should avoid passing the entire `state` object as a prop. `mapStateToProps` exists specifically to limit the scope of props we're passing to a connected component, and these condition helpers can run there.

**Comment by tough-griff** [comment]
re: re: @jonambas @jasonrhodes regarding `<Layout loading>` vs `<Loading>`, I think a `Loading` component that behaves similar to a modal (using React 16 `Portals` even!?) is not a bad pattern, but we could also trigger the `loading` prop similarly to how we handle global errors, with specific actions being fired to handle loading state.

File: docs/access-control.md (hunk 7 above)
Comment (jasonrhodes) [reply docs/access-control.md]: Yeah the only reason I needed to do it was so I could provide wholly to condition functions. I'll keep an eye on it, it's still not totally clear what the difference is between some of the ways I've seen us doing things.

File: src/components/auth/AccessControl.js (hunk 8 above)
Comment (jasonrhodes) [reply src/components/auth/AccessControl.js]: Gotcha, I think my head is getting around this, good call.
